sdk/metric/aggregator/exponential/mapping/exponent/exponent_test.go

```diff
@@ -212,11 +236,14 @@ func TestExponentMappingNegFour(t *testing.T) {
 		{0x1p-1023, -64},
```

Comment:
Some of these min / subnormal buckets appear to be incorrect. For example, I believe `0x1p-1074,` should be in the -68 bucket instead of -64.

Am I missing something or is the implementation ignoring perfect precision around subnormal values intentionally?

Reply:
I believe it is a mistake to treat subnormal values as real values.

One of the properties that I wanted to enforce in the data structure is as follows:
- When a bucket B1 is calculated at scale S1
- And a bucket B2 calculated for the same value at scale S2,
- Where S2 < S1
- We expect (B1 >> (S1-S2)) == B2, which follows from the "perfect subsetting"

The simple logarithm mapping functions are not accurate in the subnormal range. Rounding values up to the smallest normal value ensures the property above. Imagine a histogram where the first entered value is 0x1p-1050 for example. With one value, we can use the maximum scale, which is 20 in this case. However the logarithm implementation does not function for subnnormals, thus to ensure that the bucket computed at scale 20 matches the one for negative scales, we round up subnormals.

I would like to reserve the right to improve both mapping functions -- as long as the last property holds the data structure will not be impacted by the decision to support subnormals.

Reply:
That makes sense. But the [implementation from the spec](https://github.com/open-telemetry/opentelemetry-specification/pull/2633/files#diff-6c97e08ad12fe1538de64c06f0cab436e9115287ea35183b88dbd1df77ce0bc5R745-R759) has special logic accommodating subnormal values which applies to scales <= 0 since negative scales are just a [bitshift of the zero scale algorithm](https://github.com/open-telemetry/opentelemetry-specification/pull/2633/files#diff-6c97e08ad12fe1538de64c06f0cab436e9115287ea35183b88dbd1df77ce0bc5R803-R805).

Reply:
Yes. Although I wrote that (with @reyang's help 😂), I followed it with the [another example allowing implementations to round subnormal values](https://github.com/open-telemetry/opentelemetry-specification/pull/2633/files#diff-6c97e08ad12fe1538de64c06f0cab436e9115287ea35183b88dbd1df77ce0bc5R772-R790). The requirement being imposed by me here is that the two mapping functions handle subnormals the same way, which does not feel like a spec-level requirement, #3022 requires the two mapping functions to be consistent.

TL;DR Another reason to avoid treating subnormals, IMO, is that it changes the total maximum range of the data structure in addition to making it less symmetric.  Normal exponents range from [-1022, 1023], which fits nicely into 11 bits of information. If you allow subnormal values, the range is [-1074, 1023] which requires 12 bits of information (with most of 1 bit unused). Since we are limited to 32 bits of index (in OTLP), taking 12 bits of information for the index would decrease the maximum and minimum scale parameters used in #3022. MaxScale=20 implies 20 bits of index per exponent, so with only normal values MaxScale=20 plus 11 bits of exponent fits a 32 bit integer. Allowing subnormals would limit MaxScale to 19 and require MinScale=-11.

Reply:
Here are some test cases I wrote for OpenTelemetry.NET just FYI https://github.com/open-telemetry/opentelemetry-dotnet/blob/245c2b3e29a350e3cfef2168dcb29be1fcda38f8/test/OpenTelemetry.Tests/Metrics/ExponentialBucketHistogramTest.cs#L146.